nvflare/apis/utils/decomposers/flare_decomposers.py

```diff
@@ -0,0 +1,155 @@
+
+    def recompose(self, data: Any) -> Shareable:
+        obj = Shareable()
+        for k, v in data.items():
+            obj[k] = v
+        return obj
+
+
+class ContextDecomposer(Decomposer):
+    @staticmethod
+    def supported_type():
+        return FLContext
+
+    def decompose(self, target: FLContext) -> Any:
+        return [target.model, target.props]
+
+    def recompose(self, data: Any) -> FLContext:
+        obj = FLContext()
+        obj.model = data[0]
+        obj.props = data[1]
+        return obj
+
+
+class DxoDecomposer(Decomposer):
+    @staticmethod
+    def supported_type():
+        return DXO
+
+    def decompose(self, target: DXO) -> Any:
+        return [target.data_kind, target.data, target.meta]
+
+    def recompose(self, data: Any) -> DXO:
+        return DXO(data[0], data[1], data[2])
+
+
+class ClientDecomposer(Decomposer):
+    @staticmethod
+    def supported_type():
+        return Client
+
+    def decompose(self, target: Client) -> Any:
+        return [target.name, target.token, target.last_connect_time, target.props]
+
+    def recompose(self, data: Any) -> Client:
+        client = Client(data[0], data[1])
+        client.last_connect_time = data[2]
+        client.props = data[3]
+        return client
+
+
+class RunSnapshotDecomposer(Decomposer):
+    @staticmethod
+    def supported_type():
+        return RunSnapshot
+
+    def decompose(self, target: RunSnapshot) -> Any:
+        return [target.component_states, target.completed, target.job_id]
+
+    def recompose(self, data: Any) -> RunSnapshot:
+        snapshot = RunSnapshot(data[2])
+        snapshot.component_states = data[0]
+        snapshot.completed = data[1]
+        return snapshot
+
+
+class WorkspaceDecomposer(Decomposer):
+    @staticmethod
+    def supported_type():
+        return Workspace
+
+    def decompose(self, target: Workspace) -> Any:
+        return [target.root_dir, target.name, target.config_folder]
+
+    def recompose(self, data: Any) -> Workspace:
+        return Workspace(data[0], data[1], data[2])
+
+
+class SignalDecomposer(Decomposer):
+    @staticmethod
+    def supported_type():
+        return Signal
+
+    def decompose(self, target: Signal) -> Any:
+        return [target.value, target.trigger_time, target.triggered]
+
+    def recompose(self, data: Any) -> Signal:
+        signal = Signal()
+        signal.value = data[0]
+        signal.trigger_time = data[1]
+        signal.triggered = data[2]
+        return signal
+
+
+class NamespaceDecomposer(Decomposer):
+    @staticmethod
+    def supported_type():
+        return Namespace
+
+    def decompose(self, target: Namespace) -> Any:
+        return vars(target)
+
+    def recompose(self, data: Any) -> Namespace:
+        return Namespace(**data)
+
+
+def register():
+    if register.registered:
+        return
+
+    fobs.register_folder(os.path.dirname(__file__), __package__)
+    register.registered = True
+
+
+register.registered = False
```

Comment:
where is this "register" defined?

Reply:
This is a function attribute, used as a static var for the function.